open_spiel/games/othello.cc

```diff
@@ -0,0 +1,344 @@
+    int count_one = DiskCount(Player(1));
+
+    if (count_zero > count_one) {
+      outcome_ = Player(0);
+    } else if (count_zero < count_one) {
+      outcome_ = Player(1);
+    }
+  }
+
+  current_player_ = 1 - current_player_;
+  num_moves_ += 1;
+}
+
+std::vector<Action> OthelloState::LegalRegularActions(Player p) const {  // list 
+  std::vector<Action> moves;
+  for (int cell = 0; cell < kNumCells; ++cell) {
+    if (ValidAction(p, cell)) {
+      moves.push_back(cell);
+    }
+  }
+
+  return moves;
+}
+
+std::vector<Action> OthelloState::LegalActions() const {
+  if (IsTerminal()) return {};
+  
+  // can move in an empty cell that captures
+  std::vector<Action> moves = LegalRegularActions(current_player_);
+  if (moves.empty()) {
+    moves.push_back(passMove);  // pass
+  }
+
+  return moves;
+}
+
+std::string OthelloState::ActionToString(Player player,
+                                           Action action_id) const {
+
+  if (action_id == passMove) {
+    return absl::StrCat(StateToString(PlayerToState(player)), "(pass)");
+  } else {
+    return absl::StrCat(StateToString(PlayerToState(player)), "(",
+                      action_id / kNumCols, ",", action_id % kNumCols, ")");
+  }
+}
+
+bool OthelloState::IsFull() const { 
+  for (int i = 0; i < kNumCells; i++) {
+    if (board_[i] == CellState::kEmpty) return false;
+  }
+
+  return true;
+}
+
+OthelloState::OthelloState(std::shared_ptr<const Game> game) : State(game) {
+  std::fill(begin(board_), end(board_), CellState::kEmpty);
+  board_[27] = CellState::kWhite;
+  board_[28] = CellState::kBlack;
+  board_[35] = CellState::kBlack;
+  board_[36] = CellState::kWhite;
+}
+
+std::string OthelloState::ToString() const {
+  std::string str;
+  for (int r = 0; r < kNumRows; ++r) {
+    for (int c = 0; c < kNumCols; ++c) {
+      absl::StrAppend(&str, StateToString(BoardAt(r, c)));
+    }
+    if (r < (kNumRows - 1)) {
+      absl::StrAppend(&str, "\n");
+    }
+  }
+  return str;
+}
+
+bool OthelloState::IsTerminal() const {
+  return outcome_ != kInvalidPlayer || NoValidActions();
+}
+
+std::vector<double> OthelloState::Returns() const {
+  if (outcome_ == Player{0}) {
+    return {1.0, -1.0};
+  } else if (outcome_ == Player{1}) {
+    return {-1.0, 1.0};
+  } else {
+    return {0.0, 0.0};
+  }
+}
+
+std::string OthelloState::InformationStateString(Player player) const {
+  return HistoryString();
+}
+
+std::string OthelloState::ObservationString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return ToString();
+}
+
+void OthelloState::ObservationTensor(Player player,
+                                       std::vector<double>* values) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+
+  // Treat `values` as a 2-d tensor.
+  TensorView<2> view(values, {kCellStates, kNumCells}, true);
+  for (int cell = 0; cell < kNumCells; ++cell) {
+    view[{static_cast<int>(board_[cell]), cell}] = 1.0;
```

Comment:
Many learning algorithms work better if these observations are player relative instead of absolute, so the layers would be: mine, opponent, empty; instead of p1, p2, empty.

Reply:
Done. You can now pass the Player, and the board string and observation tensor will switch.